### DISCUSSION TITLE
problem execute a python script

### DISCUSSION THREAD

**Discussion by GammaLyra** [the-anylogic-company/AnyLogic-Pypeline#56] [Q&A] [answered]
hi!
I am trying to run a python script in Anylogic using the Pypeline. In the script (.py) there is a function with two input variables, that returns me a value. I want to obtain this variable in Anylogic. I have done it in the following way:

pyCommunicator.run("import name_script");

String dia_st = String.valueOf(dia);
String mes_st = String.valueOf(mes);

Attempt attempt = pyCommunicator.runFile("name_script.name_func", dia_st,mes_st);

When I execute the simulation in Anylogic, I obtain this message: "[Errno 2] No such file or directory". The script is in the same folder that the simulation...

Thanks for everything!

Best regards

**Comment by fabricio-fn** [comment]
the error is happening because the system cannot find the Python file path.

make sure you used the .p extension (file.py)
If that doesn't work, try using the absolute path.

or it may be that the other file needs to be exported

**Reply by GammaLyra** [reply]
I've already got it working. as follows:

in my script in Python:

"import argparse

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("dia", type=int)
    parser.add_argument("mes", type=int)
    args = parser.parse_args()
    dia=args.dia
    mes=args.mes
    a=prediccion(dia,mes)"

And I do the call in Anylogig like this:

"String dia_st = String.valueOf(dia);
String mes_st = String.valueOf(mes);
Attempt attempt = pyCommunicator.runFile("pred_xgboost.py", dia_st,mes_st);"

However, my python function "prediccion(dia,mes)", have a output variable that I need to use in Anylogic. How can i access it?

Thanks for everything!

**Reply by t-wolfeadam** [reply]
The only thing you're missing here is to convert the `Attempt` object to the intended data type. I'll assume the type is a `double`; there's two ways to do it depending on your preference:

With your existing style:
```java
// ...other code...
Attempt attempt = pyCommunicator.runFile("pred_xgboost.py", dia_st,mes_st);
double myValue = attempt.getFeedback(double.class);
```

Or using the shortcut (skipping the `Attempt` object):
```java
// ...other code...
double myValue = pyCommunicator.runFile(double.class, "pred_xgboost.py", dia_st, mes_st);
```

Also, I just realized `runFile` is not documented in the wiki! I'll put that on my todo :sweat_smile:

**Comment by fabricio-fn** [comment]
try using the "subprocess" module

import argparse

def prediction(day, month):
      # Perform necessary processing to obtain the result
      result = "Your result here"
      return result

if  name == "main":
      parser = argparse.ArgumentParser()
      parser.add_argument("day", type=int)
      parser.add_argument("month", type=int)
      args = parser.parse_args()
      day = args.day
      month = args.month
      prediction_result = prediction(day, month)
      print(prediction_result)  # Print the result to be captured by AnyLogic



String day_st = String.valueOf(day);
String month_st = String.valueOf(month);
String result = "";

try {
      Process process = Runtime.getRuntime().exec("python pred_xgboost.py " + day_st + " " + month_st);
      BufferedReader reader = new BufferedReader(new  InputStreamReader(process.getInputStream()));
      result = reader.readLine(); // Capture the first output line (result) from Python
      process.waitFor(); // Wait until the Python process finishes
} catch (IOException | InterruptedException e) {
      e.printStackTrace();
}

**Reply by t-wolfeadam** [reply]
Fortunately, this route is unnecessary, as it's exactly the type of thing that Pypeline as a whole (and specifically `runFile`) is intended to serve as an abstraction away from!

Also, to avoid potential confusion in the future, you may wish to encapsulate code blocks in triple brackets, and can add syntax highlighting ([github help doc ref](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks)).

**Comment by t-wolfeadam** [comment, marked as answer]
The problem is that `runFile` is only intended for non-interactive execution of scripts whereas `run` and `runResults` are interactive. Because you imported the file with `run`, you want to use `runResults` instead of `runFile`; you also don't need to convert them to (Java) string first, as the function will handle that for you. 

```java
Attempt attempt = pyCommunicator.runResults("name_script.name_func(%s, %s)", dia, mes);
```

~~Note that if `name_func` is expecting those values as (Python) strings, you'll need to put single quotes around the string formatters (%s).~~ (I noticed your other comment shows it's expecting integers, so the code above is valid)

**Reply by GammaLyra** [reply]
Thank you very much for the solution and the quick response! it works for me :) I hadn't understand well and thought that the iterative command was runFile...

Thanks for everything!
